### PR TITLE
Potential leak of handles and memory

### PR DESCRIPTION
**Pull request by snjeza**
Fixes https://github.com/redhat-developer/vscode-xml/issues/108
Related PR - https://github.com/eclipse/eclipse.jdt.ls/pull/931

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

### CODE REVIEW

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java [hunk 1]
```diff
@@ -83,14 +83,28 @@ private boolean parentProcessStillRunning() {
...
+		boolean result = false;
 		try {
-			Process process = Runtime.getRuntime().exec(command);
-			int processResult = process.waitFor();
-			return processResult == 0;
+			process = Runtime.getRuntime().exec(command);
+			result = process.waitFor(POLL_DELAY_SECS, TimeUnit.SECONDS);
+			if (!result) {
+				process.destroy();
+				result = process.waitFor(POLL_DELAY_SECS, TimeUnit.SECONDS);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: why do you wait twice?

Comment (fbricon) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: On SO, this [answer](https://stackoverflow.com/a/5483976/753170) mentions having to read the streams. Does it apply here?

**Comment by fbricon** [comment]
I'll paste @snjeza's comment for the jdt.ls fix here too:

```
Java doesn't call the finalize method when terminating a process. Because of that, when Java LS isn't used, it creates a lot of zombie processes on Windows.
See https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ParentProcessWatcher.java#L64

Steps to reproduce:

* start VS Code and Java LS, but don't use it
* download https://download.sysinternals.com/files/Handle.zip and unpack it
* call the following command several times:

```
handle64 -s -p <Java_LS_pid> | grep Process
```
You will see that the number of processes is constantly increasing.

The solution is to call System.gc(). We can also call it on Linux and Mac because it runs when Java LS isn't active.
```

**Comment by fbricon** [comment]
@snjeza did you mean `destroy` instead of `finalize`?

**Comment by fbricon** [comment]
Ok so IIUC, we need to call the GC explicitly so that it calls finalize() on the process, which in turn releases the handles, right?

**Comment by snjeza** [comment]
> Ok so IIUC, we need to call the GC explicitly so that it calls finalize() on the process, which in turn releases the handles, right?

Right.

**Comment by snjeza** [comment]
@fbricon I have updated the PR.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java [hunk 2]
```diff
@@ -83,14 +83,32 @@ private boolean parentProcessStillRunning() {
...
 			command = "ps -p " + pid;
 		}
+		Process process = null;
+		boolean result = false;
 		try {
-			Process process = Runtime.getRuntime().exec(command);
-			int processResult = process.waitFor();
-			return processResult == 0;
+			process = Runtime.getRuntime().exec(command);
+			result = process.waitFor(POLL_DELAY_SECS, TimeUnit.SECONDS);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: s/result/finished/g

Comment (snjeza) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: changed.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java (hunk 1 above)
Comment (snjeza) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: No, it doesn't.

Comment (snjeza) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: > why do you wait twice?

We wait for the process to stop.

**Comment by snjeza** [comment]
> @snjeza did you mean destroy instead of finalize?

finalize.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java [hunk 3]
```diff
@@ -83,14 +83,32 @@ private boolean parentProcessStillRunning() {
...
 		} catch (IOException | InterruptedException e) {
-			//JavaLanguageServerPlugin.logException(e.getMessage(), e);
+			// JavaLanguageServerPlugin.logException(e.getMessage(), e);
 			return true;
+		} finally {
+			if (process != null) {
+				if (!finished) {
+					process.destroyForcibly();
+				}
+				// Java doesn't call the finalize method when terminating a process.
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: So let me rephrase:

> Terminating or destroying the Process doesn't close the process handle. It is only closed when the Process object is garbage collected (in its finalize() method).
> On Windows, when the Java LS is idle, we need to explicitly request a GC, to prevent an accumulation of zombie processes, as finalize() will be called.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java [hunk 4]
```diff
@@ -83,14 +83,32 @@ private boolean parentProcessStillRunning() {
...
 			return true;
+		} finally {
+			if (process != null) {
+				if (!finished) {
+					process.destroyForcibly();
+				}
+				// Java doesn't call the finalize method when terminating a process. 
+				// Because of that, when Java LS isn't used, it creates a lot of zombie processes on Windows.
+				// GC calls finalize
+				// We can also call it on Linux and Mac because it runs when Java LS isn't active.
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: While calling the GC on an idle server is not a big deal on Linux and Mac, it will still cause unnecessary CPU usage at this point. So we might as well do it only for Windows

Comment (snjeza) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/ParentProcessWatcher.java]: I have updated the PR.

**Comment by fbricon** [comment]
@snjeza so it took me a while to test on windows because my Win 10 VM is excessively slow, but after the fix, I can see that running `handle64 -s -p <Java_LS_pid> | findstr Process` consistently returns 1, which is great, but I see that `handle64 -s -p <Java_LS_pid> | findstr File` shows file handles keep growing

**Comment by fbricon** [comment]
The total amount of handles reported by handle64 is consistent with the handles reported in the task manager. And it keeps growing with the patch anyway

**Comment by snjeza** [comment]
@fbricon checking...

**Comment by NikolasKomonen** [comment]
Looks fine to me, caps at 26 file handlers

**Comment by fbricon** [comment]
https://youtu.be/Z0fw1SmQKhI?t=52 tested on https://developer.microsoft.com/en-us/windows/downloads/virtual-machines

**Comment by snjeza** [comment]
> https://youtu.be/Z0fw1SmQKhI?t=52 tested on https://developer.microsoft.com/en-us/windows/downloads/virtual-machines

@fbricon could you test on real Windows?

**Comment by fbricon** [comment]
No I don't have the machines for that.
@yaohaizh could you check please (this PR or jdt.ls's)? I'm sure Microsoft has some Windows machines for you guys.

**Comment by yaohaizh** [comment]
@fbricon  Tested on physical Windows system, both the file & process handles numbers ( ~133 & 1 on my machine) don't increase with the patch. Looks good to me.

**Comment by fbricon** [comment]
Ok, let's merge it and see if more people can test it before we release an update.

**Comment by NikolasKomonen** [comment]
@yaohaizh Heres the link to get the latest VSIX, if you have time could you check if it works on your machine.

http://download.jboss.org/jbosstools/vscode-xml/staging/?C=M;O=D
